sdk/samples/iot/sample_pnp.c

```diff
@@ -0,0 +1,428 @@
+      {
+        if (az_failed(az_json_reader_next_token(json_reader)))
+        {
+          printf("Failed to next token\r\n");
+          return AZ_ERROR_UNEXPECTED_CHAR;
+        }
+        continue;
+      }
+
+      if (az_failed(az_json_token_get_string(
+              &(json_reader->token), (char*)scratch_buf, (int32_t)scratch_buf_len, (int32_t*)&len)))
+      {
+        printf("Failed to get string property value\r\n");
+        return AZ_ERROR_UNEXPECTED_CHAR;
+      }
+
+      if (az_failed(az_json_reader_next_token(json_reader)))
+      {
+        printf("Failed to get next token\r\n");
+        return AZ_ERROR_UNEXPECTED_CHAR;
+      }
+
+      property_callback(
+          component_name,
+          az_span_init((uint8_t*)scratch_buf, len),
+          &(json_reader->token),
+          version,
+          context_ptr);
+    }
+
+    if (json_reader->token.kind == AZ_JSON_TOKEN_BEGIN_OBJECT)
+    {
+      if (az_failed(az_json_reader_skip_children(json_reader)))
+      {
+        printf("Failed to skip children of object\r\n");
+        return AZ_ERROR_UNEXPECTED_CHAR;
+      }
+    }
+    else if (json_reader->token.kind == AZ_JSON_TOKEN_END_OBJECT)
+    {
+      break;
+    }
+  }
+
+  return AZ_OK;
+}
+
+// Move reader to the value of property name
+static az_result sample_json_child_token_move(az_json_reader* json_reader, az_span property_name)
+{
+  while (az_succeeded(az_json_reader_next_token(json_reader)))
+  {
+    if ((json_reader->token.kind == AZ_JSON_TOKEN_PROPERTY_NAME)
+        && az_json_token_is_text_equal(&(json_reader->token), property_name))
+    {
+      if (az_failed(az_json_reader_next_token(json_reader)))
+      {
+        printf("Failed to read next token\r\n");
+        return AZ_ERROR_UNEXPECTED_CHAR;
+      }
+
+      return AZ_OK;
+    }
+    else if (json_reader->token.kind == AZ_JSON_TOKEN_BEGIN_OBJECT)
+    {
+      if (az_failed(az_json_reader_skip_children(json_reader)))
+      {
+        printf("Failed to skip child of complex object\r\n");
+        return AZ_ERROR_UNEXPECTED_CHAR;
+      }
+    }
+    else if (json_reader->token.kind == AZ_JSON_TOKEN_END_OBJECT)
+    {
+      return AZ_ERROR_ITEM_NOT_FOUND;
+    }
+  }
+
+  return AZ_ERROR_ITEM_NOT_FOUND;
+}
+
+// Check if the component name is in the model
+static az_result is_component_in_model(
+    az_span component_name,
+    const az_span** sample_components_ptr,
+    int32_t sample_components_num,
+    int32_t* out_index)
+{
+  int32_t index = 0;
+
+  if (az_span_ptr(component_name) == NULL || az_span_size(component_name) == 0)
+  {
+    return AZ_ERROR_UNEXPECTED_CHAR;
+  }
+
+  while (index < sample_components_num)
+  {
+    if ((az_span_size(component_name) == az_span_size(*sample_components_ptr[index]))
+        && (memcmp(
+                (void*)az_span_ptr(component_name),
+                (void*)az_span_ptr(*sample_components_ptr[index]),
+                (size_t)az_span_size(component_name))
+            == 0))
```

Comment:
Try to limit uses of memcmp and other ptr based methods as much as possible and try to stay in the "span world" where possible.

In this case, can you use `az_span_is_content_equal`?
This is essentially re-implementing parts of the existing API, so we should avoid that.